### ISSUE TITLE
 Clang try to eagerly instantiate more things that needed

### ISSUE THREAD

**Issue by fsb4000** [llvm/llvm-project#61254]
@EricWF wrote that the code should work, but it will stop working if we change `declval` to match the C++ standard.

```c++
template <class T>
__add_rvalue_reference(T) declval() noexcept {
    static_assert(!__is_same(T, T),
                  "Calling declval is ill-formed, see [declval]/2.");
}

template <class T, class U, class = U[declval<T>()]>
constexpr bool xxx(int) {
    return true;
}

template <class T, class U>
constexpr bool xxx(long) {
    return false;
}

constexpr bool y = xxx<long long, int>(0);
```

https://gcc.godbolt.org/z/b8cxG89Mb

I've created the same issue for GCC.

**Comment by fsb4000** [comment]
Actually, I'm not sure that the code should work.

Calling `declval` is ill-formed and we call it. So Clang failed as expected.  Did I miss something?

**Comment by TheInferentialImp** [comment]
The error message is indicating that there is a static assertion failure due to the requirement !__is_same(T, T) in the declval function template.

The declval function is part of the C++11 type traits library, and is used to obtain a reference to an object of a given type without actually constructing that object. This can be useful in certain template programming scenarios where a reference to an object of a given type is needed, but that object cannot be constructed or copied.

In the code you provided, the declval function template is declared with a single parameter of type T, and it returns an rvalue reference to that type using the __add_rvalue_reference type trait. The __add_rvalue_reference type trait is used to add an rvalue reference to a given type, and it is used here to ensure that declval returns an rvalue reference.

However, the static_assert in declval is causing the error message to be generated. The __is_same type trait is used to check if two types are the same, and in this case, it is being used to check if T is the same as T. This might seem redundant, but it is actually a way to trigger a static assertion failure with a helpful error message.

The error message is indicating that the static_assert failed because the __is_same check evaluated to true, which means that T is the same as T. This error message is being generated because the declval function template is being called with a template argument of long long, which causes the static_assert to trigger.

The xxx function template is then called with the template arguments long long and int. The first overload of xxx is selected because the call to declval<long long>() fails to compile due to the static_assert in declval. The long template parameter of the first overload is then selected because the call to xxx<long long, int>(0) matches the second template parameter, which is a non-type template parameter of type U.

Overall, the error message is indicating that there is an issue with the way that declval is being used in the xxx function template. 

I think the error is being triggered because the static_assert in the declval function is being evaluated during the instantiation of the function template xxx<long long, int>(0) on line 10.

The static_assert checks that the expression !__is_same(T, T) is false, which will always be the case, since T is the same type as T. This means that the static assertion will always fail, and an error will be generated.

It seems that the intention of this code was to provide a mechanism for checking if T can be implicitly converted to U. However, the current implementation is incorrect and is triggering the static assertion error.

**Comment by nico** [comment]
Please don't spam bug trackers with chatgpt copypasta.

**Comment by TheInferentialImp** [comment]
@nico sorry about that, I should've referenced something else regarding this issue. However, to be fair, chatgpt is very helpful in many circumstances w/ bugs. Again, I apologize- just trying to help.

**Comment by ldionne** [comment]
Yeah, I think the code is ill-formed. Doesn't Clang have to instantiate `declval()` and even execute it in order to figure out the bounds of the array and whether any promotion needs to happen?
